File: modules/printer-module/ios/src/PrinterModule.swift
import ExpoModulesCore
import StarIO10

public class PrinterModule: Module, StarDeviceDiscoveryManagerDelegate {

    private var manager: StarDeviceDiscoveryManager? = nil
    private var devices: String = "Initial value"

    public func definition() -> ModuleDefinition {
        Name("PrinterModule")

        Events("onDeviceFound")

        Function("getDevices") { () -> String in
            devices
        }

        Function("onScan") { () -> Void in
            sendDeviceFoundEvent(valueToAppend: "Scan starting...")
            do {
                try manager = StarDeviceDiscoveryManagerFactory.create(
                    interfaceTypes: [InterfaceType.bluetooth])
                manager?.discoveryTime = 1000
                manager?.delegate = self

                try manager?.startDiscovery()
            } catch let error {
                print("Error \(error)")
            }
        }
    }

    public func manager(
        _ manager: any StarIO10.StarDeviceDiscoveryManager,
        didFind printer: StarIO10.StarPrinter
    ) {
        let interfaceType = printer.connectionSettings.interfaceType
        let identifier = printer.connectionSettings.identifier

        sendDeviceFoundEvent(valueToAppend: "\(interfaceType): \(identifier)")
    }

    public func managerDidFinishDiscovery(
        _ manager: any StarIO10.StarDeviceDiscoveryManager
    ) {
        sendDeviceFoundEvent(valueToAppend: "Scan finished")
    }

    public func sendDeviceFoundEvent(valueToAppend: String) {
        devices.append("\n\(valueToAppend)")
        sendEvent("onDeviceFound", ["devices": devices])
    }
}

File: modules/printer-module/android/src/main/java/expo/modules/printermodule/PrinterModule.kt
package expo.modules.printermodule

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import com.starmicronics.stario10.InterfaceType
import com.starmicronics.stario10.StarDeviceDiscoveryManager
import com.starmicronics.stario10.StarDeviceDiscoveryManagerFactory
import com.starmicronics.stario10.StarPrinter
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PrinterModule : Module() {

    private var manager: StarDeviceDiscoveryManager? = null
    private var devices: String = "Initial value"

    companion object {
        private const val TAG = "PrinterModule"
    }

    override fun definition() = ModuleDefinition {
        Name("PrinterModule")

        Events("onDeviceFound")

        Function("getDevices") {
            devices
        }

        Function("onScan") {
            sendOnDeviceFoundEvent("Scan starting...")
            if (hasBluetoothPermission()) {
                try {
                    manager?.stopDiscovery()

                    appContext.reactContext?.let { context ->
                        manager = StarDeviceDiscoveryManagerFactory.create(listOf(InterfaceType.Bluetooth), context).apply {
                            discoveryTime = 1000
                            callback = object : StarDeviceDiscoveryManager.Callback {

                                override fun onPrinterFound(printer: StarPrinter) {
                                    sendOnDeviceFoundEvent("${printer.connectionSettings.interfaceType}:${printer.connectionSettings.identifier}")
                                }

                                override fun onDiscoveryFinished() {
                                    sendOnDeviceFoundEvent("Discovery finished.")
                                }
                            }
                            startDiscovery()
                        }
                    }
                } catch (e: Exception) {
                    Log.d(TAG, "Error: $e")
                }
            } else {
                requestBluetoothPermission()
            }
        }
    }

    private fun requestBluetoothPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return
        }

        appContext.currentActivity?.apply {
            if (checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(
                    arrayOf(Manifest.permission.BLUETOOTH_CONNECT), 10000
                )
            }
        }
    }

    private fun hasBluetoothPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return true
        }

        return appContext.currentActivity?.checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED
    }

    fun sendOnDeviceFoundEvent(valueToAppend: String) {
        devices += "\n$valueToAppend"
        sendEvent("onDeviceFound", mapOf("devices" to devices))
    }
}
